src/Interpreters/Cache/FileSegment.h

```diff
@@ -305,46 +315,37 @@ friend class StorageSystemFilesystemCache;
 
     RemoteFileReaderPtr remote_file_reader;
     LocalCacheWriterPtr cache_writer;
-    bool detached_writer = false;
 
     /// downloaded_size should always be less or equal to reserved_size
-    size_t downloaded_size = 0;
-    size_t reserved_size = 0;
+    std::atomic<size_t> downloaded_size = 0;
+    std::atomic<size_t> reserved_size = 0;
 
     /// global locking order rule:
     /// 1. cache lock
     /// 2. segment lock
 
-    mutable std::mutex mutex;
+    mutable FileSegmentGuard segment_guard;
+    LockedKeyCreatorPtr locked_key_creator;
     std::condition_variable cv;
 
-    /// Protects downloaded_size access with actual write into fs.
-    /// downloaded_size is not protected by download_mutex in methods which
-    /// can never be run in parallel to FileSegment::write() method
-    /// as downloaded_size is updated only in FileSegment::write() method.
-    /// Such methods are identified by isDownloader() check at their start,
-    /// e.g. they are executed strictly by the same thread, sequentially.
-    mutable std::mutex download_mutex;
-
     Key file_key;
+    const std::string file_path;
     FileCache * cache;
 
     Poco::Logger * log;
 
     /// "detached" file segment means that it is not owned by cache ("detached" from cache).
     /// In general case, all file segments are owned by cache.
     bool is_detached = false;
```

Comment:
AFAIU, the flag is redundant since it's equivalent of `download_state == Stake::SKIP_CACHE`